src/librepgp/stream-parse.cpp

```diff
@@ -859,51 +859,40 @@ signed_read_single_signature(pgp_source_signed_param_t *param,
                              pgp_source_t *             readsrc,
                              pgp_signature_t **         sig)
 {
-    uint8_t               ptag;
-    int                   ptype;
-    pgp_signature_t       readsig;
-    pgp_signature_info_t *siginfo;
-
+    uint8_t ptag;
     if (!src_peek_eq(readsrc, &ptag, 1)) {
         RNP_LOG("failed to read signature packet header");
         return RNP_ERROR_READ;
     }
 
-    ptype = get_packet_type(ptag);
-
+    int ptype = get_packet_type(ptag);
     if (ptype != PGP_PKT_SIGNATURE) {
         RNP_LOG("unexpected packet %d", ptype);
         return RNP_ERROR_BAD_FORMAT;
     }
 
     try {
         param->siginfos.push_back({});
-    } catch (const std::exception &e) {
-        RNP_LOG("%s", e.what());
-        return RNP_ERROR_OUT_OF_MEMORY;
-    }
-    siginfo = &param->siginfos.back();
-
-    if (stream_parse_signature(readsrc, &readsig) != RNP_SUCCESS) {
-        RNP_LOG("failed to parse signature");
-        siginfo->unknown = true;
+        pgp_signature_info_t &siginfo = param->siginfos.back();
+        pgp_signature_t       readsig;
+        if (readsig.parse(*readsrc)) {
+            RNP_LOG("failed to parse signature");
+            siginfo.unknown = true;
+            if (sig) {
+                *sig = NULL;
+            }
+            return RNP_SUCCESS;
+        }
+        param->sigs.push_back(std::move(readsig));
+        siginfo.sig = &param->sigs.back();
```

Comment:
Aren't you concerned with reallocations when vector grows?

Reply:
That's why param->sigs is of `std::list` type - it doesn't invalidate pointers on addition/deletion. Got hit here with reallocations in one of the older PRs.

Reply:
Good to hear that. Somehow missed that, sorry.